### PR TITLE
Adding var-window FMHA context support for PrimsTS.

### PR DESCRIPTION
**Pull request by mingxu1067**
## 📌 Description

1. Adding variable attention window support which allows each Q token has its own window range.
2. Performance:
- Q/K/V/O shape: [64, 2560, 8, D], D=128 or 256.
- Dtype: [FP15, FP8]
- 4 Seqence in one batchL T1[0:256], I1[256:1280], T2[1280:1536], I2[1536:2560]
- Base mask is causal with `left_window=128`.
- `I1` and `I2` additionally allow `right_window=128`, capped at their own segment end.

|  D  | Dtype  | FmhaTs ms | TRT-LLM-GEN ms |
| --- | ------ | --------- | -------------- |
| 256 |  fp16  |   0.6066  |     0.6554     |
| 128 |  fp16  |   0.3538  |     0.4027     |
| 256 |  fp8   |   0.3898  |     0.4424     |
| 128 |  fp8   |   0.3361  |     0.3231     |

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

## Summary by CodeRabbit

* **New Features**
  * Added variable-window attention support for packed sequences.
  * Planning and prefill APIs now accept per-query window start and end bounds.
  * Supports configurable attention ranges with CUDA int32 boundary tensors.
  * Improved window-boundary handling across query tiles for more accurate masking.
  * Added support for longer text and image sequences with variable attention windows.

* **Bug Fixes**
  * Added validation for window bounds, incompatible mask settings, and output aliasing.
  * Paged variable-window attention now reports a clear unsupported-configuration error.

* **Tests**
  * Added correctness coverage across multiple head dimensions, sequence lengths, and boundary cases.

### CODE REVIEW

File: flashinfer/attention/prims_ts/context.py [hunk 1]
```diff
@@ -1580,8 +1636,10 @@ def plan(
         *,
         qo_indptr: Optional[torch.Tensor] = None,
         kv_indptr: Optional[torch.Tensor] = None,
-        mask_type: Literal["dense", "causal"] = "dense",
+        mask_type: Literal["dense", "causal", "variable_window"] = "dense",
         window_left: int = -1,
+        variable_window_token_starts: Optional[torch.Tensor] = None,
+        variable_window_token_ends: Optional[torch.Tensor] = None,
```
[the remark this answers is not in the document]
Comment (rosenrodt) [reply flashinfer/attention/prims_ts/context.py:1690]: +1 on this.

File: flashinfer/attention/prims_ts/kernels/fmha_context/fmha_kernel.py [hunk 2]
```diff
@@ -515,6 +516,60 @@ def __init__(
...
+        if kwargs.get("schedule") is None:
+            raise ValueError("VariableWindow domain tasks require a captured schedule")
+        super().__init__(**kwargs)
+        self._variable_window_token_starts = variable_window_token_starts
+        self._variable_window_token_ends = variable_window_token_ends
+        self._variable_window_cta_starts = variable_window_cta_starts
+        self._num_kv_tiles = num_kv_tiles
+        self._q_stride = q_stride
+        self._cta_m = cta_m
+        self._kv_n = kv_n
```
Comment (PerkzZheng) [inline flashinfer/attention/prims_ts/kernels/fmha_context/fmha_kernel.py]: I would rename it as tile_size_q and tile_size_kv. please help clean other codes if needed. Thanks!

File: flashinfer/attention/prims_ts/kernels/fmha_context/fmha_kernel.py [hunk 3]
```diff
@@ -515,6 +516,60 @@ def __init__(
...
+        variable_window_token_starts: cute.Tensor,
+        variable_window_token_ends: cute.Tensor,
+        variable_window_cta_starts: cute.Tensor,
+        num_kv_tiles: int | Int32,
+        q_stride: int | Int32,
+        cta_m: int,
+        kv_n: int,
+        seq_idx: int,
+        batch_idx: int,
+        offset: int = 0,
```
Comment (PerkzZheng) [inline flashinfer/attention/prims_ts/kernels/fmha_context/fmha_kernel.py:535]: what offset is used for here ? we can remove it if not used.

File: flashinfer/attention/prims_ts/kernels/fmha_context/fmha_kernel.py [hunk 4]
```diff
@@ -515,6 +516,60 @@ def __init__(
         )
 
 
+class VariableWindowDomainTask(Task):
+    """Task whose K-loop domain comes from packed-Q window endpoints."""
+
+    def __init__(
```
Comment (PerkzZheng) [inline flashinfer/attention/prims_ts/kernels/fmha_context/fmha_kernel.py:524]: it would be good if you can add more comments about what each parameter means here.

File: flashinfer/attention/prims_ts/kernels/fmha_context/fmha_resources.py [hunk 5]
```diff
@@ -106,6 +107,57 @@
...
+    valid_bits: Int32,
+    score0: Float32,
+    score1: Float32,
+    score2: Float32,
+    score3: Float32,
+) -> tuple[Float32, Float32, Float32, Float32]:
+    """Expand four bitmap bits with setp and replace invalid scores.
+
+    Use and/setp because CUDA 13.3 libNVVM fails to compile the equivalent
+    `r2p.b32 {valid0, valid1, valid2, valid3}.reverse, {$r0}.b0, 0x0f;`.
```
Comment (PerkzZheng) [inline flashinfer/attention/prims_ts/kernels/fmha_context/fmha_resources.py]: we cannot expose r2p publicly IIUC.

File: flashinfer/attention/prims_ts/kernels/fmha_context/fmha_resources.py [hunk 6]
```diff
@@ -2907,6 +3045,61 @@ def packed_dense_k_masked_row_max(
...
+                Int32(tmem_x),
+            )
+            valid_bits = _bmsk_clamp(chunk_left, chunk_right - chunk_left)
+            chunk = s_data[chunk_idx]
+            masked_scores = []
+            for quad_idx in cutlass.range_constexpr(tmem_x // 4):
+                quad_base = quad_idx * 4
+                # CUDA 13.3 NVVM rejects r2p in this kernel, so expand each
+                # four-bit group with ordinary setp instructions.
+                masked_scores.extend(
```
Comment (PerkzZheng) [inline flashinfer/attention/prims_ts/kernels/fmha_context/fmha_resources.py:3087]: we can skip masking for dense tiles.

**Comment by PerkzZheng** [comment]
Also, do you know why prims_ts has better performance compared to trtllm-gen ? just want to know the optimization tricks. Thanks!

File: flashinfer/attention/prims_ts/kernels/fmha_context/fmha_kernel.py (hunk 2 above)
Comment (mingxu1067) [reply flashinfer/attention/prims_ts/kernels/fmha_context/fmha_kernel.py]: Done.

File: flashinfer/attention/prims_ts/kernels/fmha_context/fmha_kernel.py (hunk 3 above)
Comment (mingxu1067) [reply flashinfer/attention/prims_ts/kernels/fmha_context/fmha_kernel.py:535]: It is used to derive two task-domain lengths from the same variable-window K-tile span, like.
```
domain_n_kwargs = {**base_kwargs, "offset": 0}
domain_n_minus_1_kwargs = {**base_kwargs, "offset": 1}
...
last_k_tile - first_k_tile - self._offset
```
Refer to [fmha_kernel.py#L591](https://github.com/mingxu1067/flashinfer/blob/mingh/prim_ts_variable_fmha_context/flashinfer/attention/prims_ts/kernels/fmha_context/fmha_kernel.py#L591)

File: flashinfer/attention/prims_ts/kernels/fmha_context/fmha_kernel.py (hunk 4 above)
Comment (mingxu1067) [reply flashinfer/attention/prims_ts/kernels/fmha_context/fmha_kernel.py:524]: Added.

File: flashinfer/attention/prims_ts/kernels/fmha_context/fmha_resources.py (hunk 6 above)
Comment (mingxu1067) [reply flashinfer/attention/prims_ts/kernels/fmha_context/fmha_resources.py:3087]: I tried few ways to skip it, like conditional around `_mask_score_quad`, but got slightly perf regression. The `variable_window_token_starts/ends` are CUDA tensors, therefore we cannot decide the skip at compiling time.

**Comment by mingxu1067** [comment]
> Also, do you know why prims_ts has better performance compared to trtllm-gen ? just want to know the optimization tricks. Thanks!

I found there is a benchmark mismatch. TRT-LLM-Gen was using pagedKV but Prims-TS using continuous KV. By fixing this, the performance reduce from 20% to around 5-10%. I updated the perf table of this PR as well.
The remaining difference should come from that Prims-TS loads `window-start/end` outside K-loop and but TRT-LLM-Gen loads them inside K-loop.

File: flashinfer/attention/prims_ts/kernels/fmha_context/fmha_resources.py (hunk 5 above)
Comment (mingxu1067) [reply flashinfer/attention/prims_ts/kernels/fmha_context/fmha_resources.py]: Removed.

**Comment by PerkzZheng** [comment]
> > Also, do you know why prims_ts has better performance compared to trtllm-gen ? just want to know the optimization tricks. Thanks!
> 
> I found there is a benchmark mismatch. TRT-LLM-Gen was using pagedKV but Prims-TS using continuous KV. By fixing this, the performance reduce from 20% to around 5-10%. I updated the perf table of this PR as well. The remaining difference should come from that Prims-TS loads `window-start/end` outside K-loop and but TRT-LLM-Gen loads them inside K-loop.

thanks. LGTM

File: flashinfer/attention/prims_ts/kernels/fmha_context/fmha_resources.py (hunk 6 above)
Comment (PerkzZheng) [reply flashinfer/attention/prims_ts/kernels/fmha_context/fmha_resources.py:3087]: okay. it looks good to me. Previously, we assume window_start/end should always increase as offset_q so the boundary can be determined by the first token or last token. Anyway, it is something we can revisit later.

**Review by PerkzZheng — APPROVED** [review]

**Review by saltyminty — APPROVED** [review]
LGTM (mainly checked plumbing and didn't look too closely at kernel changes since Perkz already approved)

**Comment by qsang-nv** [comment]
**`cache_variable_window_bounds` reads the bound tensors out of bounds for padded Q rows.**

In `fmha_resources.py` it computes `packed_q = batch_coord * q_stride + local_q` and indexes the flattened `[B * Sq]` tensors without clamping `local_q` to `Sq - 1`. `VariableWindowDomainTask.get_domain` does clamp the same coordinate (`last_local_q = min(..., q_stride - 1)`), and `_build_variable_window_cta_starts` pads its reduction with `_INT32_MAX`, so the padded tail is handled everywhere except here.

Any `Sq` that is not a multiple of the Q CTA tile (256 for D128, 128 for D256) leaves padding rows in the last Q tile, and for the last batch `packed_q` runs past the allocation: with `B=2, Sq=33, head_dim=128` it reaches `1 * 33 + 255 = 288` while the tensors hold 66 elements. These rows do not normally affect valid outputs when the kernel completes, because their O stores are dropped by the TMA bounds. However, the accesses remain undefined and may be reported by compute-sanitizer or fault near an allocation boundary. Both new tests use tile-aligned `Sq` (2560, 256), so the path is uncovered.

Suggest clamping `local_q` as `get_domain` does, plus a regression case with unaligned `Sq` and `B > 1` (batch offsets are currently untested).

**Comment by mingxu1067** [comment]
> **`cache_variable_window_bounds` reads the bound tensors out of bounds for padded Q rows.**
> 
> In `fmha_resources.py` it computes `packed_q = batch_coord * q_stride + local_q` and indexes the flattened `[B * Sq]` tensors without clamping `local_q` to `Sq - 1`. `VariableWindowDomainTask.get_domain` does clamp the same coordinate (`last_local_q = min(..., q_stride - 1)`), and `_build_variable_window_cta_starts` pads its reduction with `_INT32_MAX`, so the padded tail is handled everywhere except here.
> 
> Any `Sq` that is not a multiple of the Q CTA tile (256 for D128, 128 for D256) leaves padding rows in the last Q tile, and for the last batch `packed_q` runs past the allocation: with `B=2, Sq=33, head_dim=128` it reaches `1 * 33 + 255 = 288` while the tensors hold 66 elements. These rows do not normally affect valid outputs when the kernel completes, because their O stores are dropped by the TMA bounds. However, the accesses remain undefined and may be reported by compute-sanitizer or fault near an allocation boundary. Both new tests use tile-aligned `Sq` (2560, 256), so the path is uncovered.
> 
> Suggest clamping `local_q` as `get_domain` does, plus a regression case with unaligned `Sq` and `B > 1` (batch offsets are currently untested).

Added

**Review by qsang-nv — APPROVED** [review]
